### PR TITLE
usbus/hid_io: add missing header file, add RX callback function

### PR DESCRIPTION
**Pull request by Ollrogge**
### Contribution description

This PR adds a header file for `sys/usb/usbus/hid/hid.io.c` which was missing before. Furthermore it adds the possibility to add a RX callback function which will be called whenever data has been received from the USB HID endpoint. This is supposed to enable the usage of the `hid_io` interface in an async context (e.g. trigger an event when data has been received).

### Testing procedure
I have extended `test/usbus_hid` to also test the functionality of the RX callback function.

### CODE REVIEW

**Review by PeterKietzmann — COMMENTED** [review]
The changes seem fine. Still works as expected (see below). 

<details>
  <summary>host</summary>

```
[67577.318364] usb 5-3: new full-speed USB device number 6 using xhci_hcd
[67599.380319] usb 5-3: new full-speed USB device number 7 using xhci_hcd
[67599.530359] usb 5-3: New USB device found, idVendor=1209, idProduct=7d01, bcdDevice= 0.00
[67599.530364] usb 5-3: New USB device strings: Mfr=3, Product=2, SerialNumber=4
[67599.530368] usb 5-3: Product: USB device
[67599.530371] usb 5-3: Manufacturer: RIOT-os.org
[67599.530373] usb 5-3: SerialNumber: 6B64081122F6AFDA
[67599.538009] hid-generic 0003:1209:7D01.001F: hiddev3,hidraw7: USB HID v1.10 Device [RIOT-os.org USB device] on usb-0000:0d:00.0-3/input0
abc@def:~$ echo "Katze" > /dev/hidraw7

```
</details>

<details>
  <summary>tests/usbus_hid on nrf52840dk</summary>

```
2021-07-29 10:25:36,370 # RIOT USB HID echo test
2021-07-29 10:25:36,377 # Write input to the hidraw device under /dev/hidrawX, and see if it gets echoed here
2021-07-29 10:25:38,405 # main(): This is RIOT! (Version: 2021.01-devel-4060-g8d5d3b-HEAD)
2021-07-29 10:25:41,387 # RIOT USB HID echo test
2021-07-29 10:25:41,394 # Write input to the hidraw device under /dev/hidrawX, and see if it gets echoed here
2021-07-29 10:26:37,302 # USB HID rx cb called
2021-07-29 10:26:37,305 # Msg received via USB HID: 
2021-07-29 10:26:37,305 # Katze
2021-07-29 10:26:37,305 # 
2021-07-29 10:30:57,439 # USB HID rx cb called
2021-07-29 10:30:57,441 # Msg received via USB HID: 
2021-07-29 10:30:57,442 # Katze

```
</details>

**Comment by PeterKietzmann** [comment]
@dylad, @bergzand do you want to give this PR a quick look? Otherwise, I would approve and merge soon.

**Comment by Ollrogge** [comment]
As of now only one callback function can be registered and would be overwritten by a subsequent call to `usb_hid_io_set_rx_cb`. Since it could be the case that multiple applications want to register this callback, a better solution would be to use a linked_list of callback functions.

What do you guys think @PeterKietzmann @dylad @bergzand ?

**Comment by fjmolinas** [comment]
> @dylad, @bergzand do you want to give this PR a quick look? Otherwise, I would approve and merge soon.

@bergzand at least is AFK (vacation)

File: sys/usb/usbus/hid/hid.c [hunk 1]
```diff
@@ -107,9 +107,9 @@ static void _init(usbus_t *usbus, usbus_handler_t *handler)
     hid->hid_descr.arg = hid;
 
     /*
-    Configure Interface as USB_HID interface, choosing NONE for subclass and
-    protocol in order to represent a generic I/O device
-    */
+       Configure Interface as USB_HID interface, choosing NONE for subclass and
+       protocol in order to represent a generic I/O device
+     */
```
Comment (fjmolinas) [inline sys/usb/usbus/hid/hid.c:112]: Are these changes a result of uncrustifying? Can you split it into a separate commit? It's hard to follow the actual changes otherwise. Same for all other occurences.

**Comment by dylad** [comment]
I am also on vacation right now.
If you think this PR is ready to go then go ahead.
Otherwise I can have a look in about two weeks

File: sys/usb/usbus/hid/hid.c (hunk 1 above)
Comment (Ollrogge) [reply sys/usb/usbus/hid/hid.c:112]: Sure thing :) Done.

Comment (fjmolinas) [reply sys/usb/usbus/hid/hid.c:112]: Thanks!

**Comment by fjmolinas** [comment]
> As of now only one callback function can be registered and would be overwritten by a subsequent call to `usb_hid_io_set_rx_cb`. Since it could be the case that multiple applications want to register this callback, a better solution would be to use a linked_list of callback functions.
> 
> What do you guys think @PeterKietzmann @dylad @bergzand ?

Can you expand on this? Do you have a use case in mind with two interesting applications? I'm not overly familiar with this, but if the HID device is a mouse wouldn't there be a single controller for all the events, and therefore only the need to register a single callback?

Wouldn't it make sense for the callback to at least have a void pointer argument?

**Comment by Ollrogge** [comment]
> > As of now only one callback function can be registered and would be overwritten by a subsequent call to `usb_hid_io_set_rx_cb`. Since it could be the case that multiple applications want to register this callback, a better solution would be to use a linked_list of callback functions.
> > What do you guys think @PeterKietzmann @dylad @bergzand ?
> 
> Can you expand on this? Do you have a use case in mind with two interesting applications? I'm not overly familiar with this, but if the HID device is a mouse wouldn't there be a single controller for all the events, and therefore only the need to register a single callback?
> 
> Wouldn't it make sense for the callback to at least have a void pointer argument?

The use case I thought about was having more than one HID device, while both want to to use `hid_io.c`. However in this case one would also need individual buffers etc. This is something to add in further PR's I think. Therefore 1 static callback function is actually fine for now I think.

Why would you add a void pointer argument ? What would you pass as argument from the USB HID layer ?

**Comment by fjmolinas** [comment]
> > > As of now only one callback function can be registered and would be overwritten by a subsequent call to `usb_hid_io_set_rx_cb`. Since it could be the case that multiple applications want to register this callback, a better solution would be to use a linked_list of callback functions.
> > > What do you guys think @PeterKietzmann @dylad @bergzand ?
> > 
> > 
> > Can you expand on this? Do you have a use case in mind with two interesting applications? I'm not overly familiar with this, but if the HID device is a mouse wouldn't there be a single controller for all the events, and therefore only the need to register a single callback?
> > Wouldn't it make sense for the callback to at least have a void pointer argument?
> 
> The use case I thought about was having more than one HID device, while both want to to use `hid_io.c`. However in this case one would also need individual buffers etc. This is something to add in further PR's I think. Therefore 1 static callback function is actually fine for now I guess.
> 
> Why would you add a void pointer argument ? What would you pass as argument from the USB HID layer ?

It wouldn't be from the USB HID layer, it would be from the upper layer, you could pass a descriptor, pointer to a structure with whatever state data, just like you do for xtimer callbacks and other. But I do not have a use case in mind.

**Comment by PeterKietzmann** [comment]
> This is something to add in further PR's I think. Therefore 1 static callback function is actually fine for now I guess.

I agree.

> Wouldn't it make sense for the callback to at least have a void pointer argument?

> What would you pass as argument from the USB HID layer ?

> It wouldn't be from the USB HID layer, it would be from the upper layer,

Generally,  the option of an argument seems like a good idea, although I don't have a specific case in mind either. @Ollrogge what is the typical use of HID IO? Can't you think of a situation where an upper layer needs to pass information to the callback? In worst case, check the API of an other implementation (for example in [KEIL](https://www.keil.com/pack/doc/mw/USB/html/_u_s_b__device.html))

**Comment by Ollrogge** [comment]
> Generally, the option of an argument seems like a good idea, although I don't have a specific case in mind either. @Ollrogge what is the typical use of HID IO? Can't you think of a situation where an upper layer needs to pass information to the callback? In worst case, check the API of an other implementation (for example in [KEIL](https://www.keil.com/pack/doc/mw/USB/html/_u_s_b__device.html))

Typical usages of HID IO would be movements of a mouse or keypresses. But since USB HID is a self-explaining protocol it could be anything (e.g. a FIDO2 authenticator as in my [PR](https://github.com/RIOT-OS/RIOT/pull/16489)). For my use case I just need the callback to trigger an event to fetch the data from the USB layer but there surely are use cases that need an arg. I think adding it doesn't hurt.

**Comment by PeterKietzmann** [comment]
Looks good. Pleas squash the third commit to the first one. I'll run a quick test afterwards

**Comment by Ollrogge** [comment]
> Looks good. Pleas squash the third commit to the first one. I'll run a quick test afterwards

Done :)

File: sys/include/usb/usbus/hid_io.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+void usb_hid_io_write(const void *buffer, size_t len);
+
+/**
+ * @brief Read data from USB HID OUT endpoint (blocking)
+ *
+ * Wrapper for @ref isrpipe_read
+ *
+ * @param[out]      buffer  buffer to read data into
+ * @param[in]       len     length of buffer
+ */
```
Comment (leandrolanzieri) [inline sys/include/usb/usbus/hid_io.h:69]:
```suggestion
 *
 * @pre `buffer != NULL`
 *
 * @param[out]      buffer  buffer to read data into
 * @param[in]       len     length of buffer
 *
 * @return Number of read bytes
 */
```

Comment (leandrolanzieri) [reply sys/include/usb/usbus/hid_io.h:69]: Please check if there are possible error return values, in that case we can add them with `@retval` bellow.

File: sys/include/usb/usbus/hid_io.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+int usb_hid_io_read(void *buffer, size_t len);
+
+/**
+ * @brief Read data from USB HID OUT endpoint (with timeout, blocking)
+ *
+ * Wrapper for @ref isrpipe_read_timeout
+ *
+ * @param[out]      buffer  buffer to read data into
+ * @param[in]       len     length of buffer
+ * @param[in]       timeout timeout in microseconds
```
Comment (leandrolanzieri) [inline sys/include/usb/usbus/hid_io.h:79]: Same here

File: sys/include/usb/usbus/hid_io.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+int usb_hid_io_read_timeout(void *buffer, size_t len, uint32_t timeout);
+
+/**
+ * @brief Initialize an USB HID IO interface
+ *
+ * @param[in]   usbus               USBUS context
+ * @param[in]   report_desc         USB_HID report descriptor
+ * @param[in]   report_desc_size    size of USB_HID report descriptor
+ */
+void usb_hid_io_init(usbus_t *usbus, uint8_t *report_desc, size_t report_desc_size);
```
Comment (leandrolanzieri) [inline sys/include/usb/usbus/hid_io.h:90]: Looks like we can make the descriptor `const`.
```suggestion
void usb_hid_io_init(usbus_t *usbus, const uint8_t *report_desc, size_t report_desc_size);
```

File: sys/usb/usbus/hid/hid_io.c [hunk 5]
```diff
@@ -25,27 +24,29 @@
...
 static usbus_hid_device_t hid;
 static uint8_t _hid_rx_buf_mem[CONFIG_USBUS_HID_INTERRUPT_EP_SIZE];
 static isrpipe_t _hid_stdio_isrpipe = ISRPIPE_INIT(_hid_rx_buf_mem);
+static usb_hid_io_t *_rx_cb;
 
-int usb_hid_io_read(void *buffer, size_t size)
+int usb_hid_io_read(void *buffer, size_t len)
 {
-    return isrpipe_read(&_hid_stdio_isrpipe, buffer, size);
+    return isrpipe_read(&_hid_stdio_isrpipe, buffer, len);
```
Comment (leandrolanzieri) [inline sys/usb/usbus/hid/hid_io.c:39]: I think here and bellow we should also `assert(buffer);` at least.

File: sys/usb/usbus/hid/hid_io.c [hunk 6]
```diff
@@ -73,10 +74,19 @@ static void _hid_rx_pipe(usbus_hid_device_t *hid, uint8_t *data, size_t len)
...
+    }
 }
 
 void usb_hid_io_init(usbus_t *usbus, uint8_t *report_desc,
                      size_t report_desc_size)
 {
     usbus_hid_init(usbus, &hid, _hid_rx_pipe,  report_desc, report_desc_size);
 }
+
+void usb_hid_io_set_rx_cb(usb_hid_io_t *usb_hid_io)
```
Comment (leandrolanzieri) [inline sys/usb/usbus/hid/hid_io.c:89]: Is there an advantage on keeping a reference to this structure, that basically holds the arguments here? Why not have:
```suggestion
void usb_hid_io_set_rx_cb(usb_hid_io_cb_t cb, void *arg)
```

File: sys/include/usb/usbus/hid_io.h [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ * @brief Set USBUS HID IO RX callback
+ *
+ * @param[in]   usb_hid_io  usb_hid_io structure to use
+ */
+void usb_hid_io_set_rx_cb(usb_hid_io_t* usb_hid_io);
+
+/**
+ * @brief Write data to USB HID IN endpoint buffer
+ *
```
[the remark this answers is not in the document]
Comment (Ollrogge) [reply sys/include/usb/usbus/hid_io.h:56]: Just a general question: Why only check that buffer is null ? If I add this shouldn't I also add e.g. `assert(usbus)` in `usb_hid_io_init`? I gues any public function that gets passed a pointer should chechk that it is != NULL ?

File: sys/usb/usbus/hid/hid_io.c (hunk 6 above)
Comment (leandrolanzieri) [reply sys/usb/usbus/hid/hid_io.c:89]: > Also I think it is a good preparation for future PR's where `hid_io.c` might have have to manage multiple HID devices at once.

I see, but then it looks more like `usb_hid_iot_t` should be used in `usb_hid_io_init`, and this function would be something like `void usb_hid_io_set_rx_cb(usb_hid_io_t *usb_hid_io, usb_hid_io_cb_t cb, void *arg)`, right? Unless I misunderstood.

Comment (Ollrogge) [reply sys/usb/usbus/hid/hid_io.c:89]: No. `usb_hid_io_init` is called to initialize the USB layer. I want the callback functionality to be an additional, not necessarily required, functionality. Possibly the naming of the struct is bad as it suggests that this struct is supposed to be used in a more genral context than just for the callback functionality.

Comment (leandrolanzieri) [reply sys/usb/usbus/hid/hid_io.c:94]: > I want the callback functionality to be an additional, not necessarily required, functionality

Agreed.

> Possibly the naming of the struct `usb_hid_io_t` is bad as it suggests that this struct is supposed to be used in a more genral context than just for the callback functionality.

Also agreed. But if this module will be a singleton and not allow multiple instantiations, the pattern used by xtimer is not necessary. I don't see the need to force the user to allocate such a structure, when it could be instantiated statically by the module (or some other `usb_hid_..` instance in the future).

Comment (Ollrogge) [reply sys/usb/usbus/hid/hid_io.c:94]: Once multiple HID reports are supported by the USB HID layer, this module will no longer be a singleton. Therefore I thought it would be a good idea to introduce a structure for this right now.

Comment (Ollrogge) [reply sys/usb/usbus/hid/hid_io.c]: Yeah I see your point. Removed this struct and adjusted it to also simply use a static arg for now

**Review by PeterKietzmann — APPROVED** [review]
Test passes

<details>
  <summary>host</summary>

```
[ 6219.135196] usb 5-3: new full-speed USB device number 4 using xhci_hcd
[ 6219.284854] usb 5-3: New USB device found, idVendor=1209, idProduct=7d01, bcdDevice= 0.00
[ 6219.284857] usb 5-3: New USB device strings: Mfr=3, Product=2, SerialNumber=4
[ 6219.284858] usb 5-3: Product: USB device
[ 6219.284859] usb 5-3: Manufacturer: RIOT-os.org
[ 6219.284860] usb 5-3: SerialNumber: 6B64081122F6AFDA
[ 6219.291421] hid-generic 0003:1209:7D01.0009: hiddev3,hidraw7: USB HID v1.10 Device [RIOT-os.org USB device] on usb-0000:0d:00.0-3/input0
abc@def:/RIOT$ echo "Hanswurst" > /dev/hidraw7


```
</details>

<details>
  <summary>tests/usbus_hid on nrf52840dk</summary>

```
2021-08-04 15:05:37,602 # main(): This is RIOT! (Version: 2021.01-devel-4061-g2c2eb-HEAD)
2021-08-04 15:05:40,582 # RIOT USB HID echo test
2021-08-04 15:05:40,589 # Write input to the hidraw device under /dev/hidrawX, and see if it gets echoed here
2021-08-04 15:05:44,348 # USB_HID rx_cb: Test argument 
2021-08-04 15:05:44,351 # Msg received via USB HID: 
2021-08-04 15:05:44,352 # Hanswurst
2021-08-04 15:05:44,352 # 

```
</details>
